tools/cassandra_delete_range/internal/cass/cass.go

```diff
@@ -308,54 +410,167 @@ func (c *ClioCass) prepareSimpleDeleteQueries(
 	fromLedgerIdx uint64,
 	toLedgerIdx uint64,
 	deleteQueryTemplate string,
-) deleteInfo {
+	colSettings columnSettings,
+) (uint64, uint64) {
+	var totalDeletes uint64
+	var totalErrors uint64
+
 	var info = deleteInfo{Query: deleteQueryTemplate}
 
-	for i := fromLedgerIdx; i <= toLedgerIdx; i++ {
-		info.Data = append(info.Data, deleteParams{Seq: i})
+	if session, err := c.clusterConfig.CreateSession(); err == nil {
+		defer session.Close()
+		for i := fromLedgerIdx; i <= toLedgerIdx; i++ {
+			info.Data = append(info.Data, deleteParams{Seq: i})
+			// for every 1000 queries in data, delete
+			if len(info.Data) == 1000 {
+				_, err := c.performDeleteQueries(&info, session, colSettings)
+				atomic.AddUint64(&totalDeletes, uint64(len(info.Data)))
+				atomic.AddUint64(&totalErrors, err)
+				info = deleteInfo{Query: deleteQueryTemplate}
+			}
+		}
+		// delete the rest of queries if exists
+		if len(info.Data) > 0 {
+			_, err := c.performDeleteQueries(&info, session, colSettings)
+			atomic.AddUint64(&totalDeletes, uint64(len(info.Data)))
+			atomic.AddUint64(&totalErrors, err)
+			info = deleteInfo{Query: deleteQueryTemplate}
+		}
+	} else {
+		log.Printf("ERROR: %s\n", err)
+		fmt.Fprintf(os.Stderr, "FAILED TO CREATE SESSION: %s\n", err)
+		atomic.AddUint64(&totalErrors, 1)
 	}
+	return totalDeletes, totalErrors
+}
 
-	return info
+func (c *ClioCass) prepareDefaultDelete(
+	scanner gocql.Scanner,
+	info *deleteInfo,
+	fromLedgerIdx maybe.Maybe[uint64],
+	toLedgerIdx maybe.Maybe[uint64],
+	rowsRetrieved *uint64,
+) bool {
+	for scanner.Next() {
+		var key []byte
+		var seq uint64
+
+		err := scanner.Scan(&key, &seq)
+		if err == nil {
+			*rowsRetrieved++
+
+			// only grab the rows that are in the correct range of sequence numbers
+			if fromLedgerIdx.HasValue() && fromLedgerIdx.Value() <= seq {
+				info.Data = append(info.Data, deleteParams{Seq: seq, Blob: key})
+			} else if toLedgerIdx.HasValue() && seq <= toLedgerIdx.Value() {
+				info.Data = append(info.Data, deleteParams{Seq: seq, Blob: key})
+			}
+		} else {
+			return false
+		}
+	}
+	return true
+}
+
+func (c *ClioCass) prepareObjectDelete(
+	scanner gocql.Scanner,
+	info *deleteInfo,
+	fromLedgerIdx maybe.Maybe[uint64],
+	toLedgerIdx maybe.Maybe[uint64],
+	rowsRetrieved *uint64,
+) bool {
+	var previousKey []byte
+	var foundLastValid bool
+	var keepLastValid = true
+
+	for scanner.Next() {
+		var key []byte
+		var seq uint64
+
+		err := scanner.Scan(&key, &seq)
+		if err == nil {
+			*rowsRetrieved++
+
+			if keepLastValid && !slices.Equal(previousKey, key) {
+				previousKey = key
+				foundLastValid = false
+			}
+
+			// only grab the rows that are in the correct range of sequence numbers
+			if fromLedgerIdx.HasValue() && fromLedgerIdx.Value() <= seq {
+				info.Data = append(info.Data, deleteParams{Seq: seq, Blob: key})
+			} else if toLedgerIdx.HasValue() {
+				if seq <= toLedgerIdx.Value() && (!keepLastValid || foundLastValid) {
```

Comment:
I saw you changed the "<" to "<=". Is it for bug fixing?

Reply:
Yes, in the initial script, I tested and there was a 1 off bug. (so if i delete before 100, it doesn't delete 99). This is because we already did -1 [here](https://github.com/XRPLF/clio/blob/2c18fd5465cfc3139a390a7ac834efdd803f9bcc/tools/cassandra_delete_range/internal/cass/cass.go#L80) so It has to be "<="